README.md
Tool to perform an action when a user scrolls in a web page.

# Installation

```bash
$ yarn add actonscroll
```

# Usage

```javascript
import Actonscroll from 'actonscroll';

// Passing arguments to the creator
(new Actonscroll({ action: () => console.log('Scrooolling!') }))
  .start();

// Calling specific method for each option
(new Actonscroll())
  .action(() => console.log('Scrooolling!'))
  .start();

// Stop
const actonscroll = new Actonscroll();
actonscroll.start();
actonscroll.stop();
```

# Documentation

<a name="Actonscroll"></a>

## Actonscroll
Executes a predefined action when the scroll event is triggered.

**Kind**: global class

* [Actonscroll](#Actonscroll)
    * [.container(container)](#Actonscroll+container) ⇒ [<code>Actonscroll</code>](#Actonscroll)
    * [.action(action)](#Actonscroll+action) ⇒ [<code>Actonscroll</code>](#Actonscroll)
    * [.conditions(conditions)](#Actonscroll+conditions) ⇒ [<code>Actonscroll</code>](#Actonscroll)
    * [.throttling(throttling)](#Actonscroll+throttling) ⇒ [<code>Actonscroll</code>](#Actonscroll)
    * [.once([once])](#Actonscroll+once) ⇒ [<code>Actonscroll</code>](#Actonscroll)
    * [.start()](#Actonscroll+start)
    * [.stop()](#Actonscroll+stop)

### new Actonscroll([options])

| Param | Type |
| --- | --- |
| [options] | <code>Object</code> |
| [options.container] | <code>Document</code> / <code>Element</code> |
| [options.action] | <code>function</code> |
| [options.conditions] | <code>Object</code> |
| [options.throttling] | <code>number</code> |
| [options.once] | <code>boolean</code> |

<a name="Actonscroll+container"></a>

### actonscroll.container(container) ⇒ [<code>Actonscroll</code>](#Actonscroll)
Sets the container element within which the scroll will be listened.

**Kind**: instance method of [<code>Actonscroll</code>](#Actonscroll)

| Param | Type |
| --- | --- |
| container | <code>Element</code> |

**Example**
```js
actonscroll.container(document.querySelector('.my-container'));
```
<a name="Actonscroll+action"></a>

### actonscroll.action(action) ⇒ [<code>Actonscroll</code>](#Actonscroll)
Sets an action to be performed when the scroll event is triggered.

**Kind**: instance method of [<code>Actonscroll</code>](#Actonscroll)

| Param | Type |
| --- | --- |
| action | <code>function</code> |

**Example**
```js
actonscroll.action(() => console.log('Scrooolling!'));
```
<a name="Actonscroll+conditions"></a>

### actonscroll.conditions(conditions) ⇒ [<code>Actonscroll</code>](#Actonscroll)
Sets a list of conditions to determine whether the configured action has to be performed or not.

**Kind**: instance method of [<code>Actonscroll</code>](#Actonscroll)

| Param | Type | Description |
| --- | --- | --- |
| conditions | <code>Object</code> |  |
| conditions.directions | <code>Array.&lt;string&gt;</code> | Allowed values: <code>'all'</code>, <code>'vertical'</code>, <code>'horizontal'</code>, <code>'up'</code>, <code>'down'</code>, <code>'left'</code>, <code>'right'</code>. |
| conditions.offset | <code>Object</code> | <code>{ x, y }</code> (In pixels). |
| conditions.custom | <code>function</code> |  |

**Example**
```js
actonscroll.conditions({
  directions: ['up', 'left'],
  offset: { y: 200 },
  custom: () => true,
});
```
<a name="Actonscroll+throttling"></a>

### actonscroll.throttling(throttling) ⇒ [<code>Actonscroll</code>](#Actonscroll)
Sets a throttling time (ms) to the scroll event.

**Kind**: instance method of [<code>Actonscroll</code>](#Actonscroll)

| Param | Type |
| --- | --- |
| throttling | <code>number</code> |

**Example**
```js
actonscroll.throttling(1000);
```
<a name="Actonscroll+once"></a>

### actonscroll.once([once]) ⇒ [<code>Actonscroll</code>](#Actonscroll)
Determines whether the action should be performed once or not.

**Kind**: instance method of [<code>Actonscroll</code>](#Actonscroll)

| Param | Type | Default |
| --- | --- | --- |
| [once] | <code>boolean</code> | <code>true</code> |

**Example**
```js
actonscroll.once();
```
<a name="Actonscroll+start"></a>

### actonscroll.start()
Executes the configured action after checking that all the conditions are satisfied.

**Kind**: instance method of [<code>Actonscroll</code>](#Actonscroll)
**Example**
```js
actonscroll.start();
```
<a name="Actonscroll+stop"></a>

### actonscroll.stop()
Removes the actonscroll from the configured container.

**Kind**: instance method of [<code>Actonscroll</code>](#Actonscroll)
**Example**
```js
actonscroll.stop();
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.2] - 2020-04-08
### Changed
- Build script in package.json to not generate sourcemaps

## [0.1.1] - 2020-04-06
### Added
- Code configurator page: https://socunanena.github.io/actonscroll/
- Changelog

## 0.1.0 - 2020-04-02
### Added
- Tool implementation
  - Container
  - Action
  - Conditions
  - Throttling
  - One time execution
- Tests implementation
- Readme

[Unreleased]: https://github.com/socunanena/actonscroll/compare/v0.1.2...HEAD
[0.1.2]: https://github.com/socunanena/actonscroll/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/socunanena/actonscroll/compare/v0.1.0...v0.1.1


src/index.js
import throttle from 'lodash.throttle';
import getScrollPosition from './helpers/getScrollPosition';
import directionsConversion from './config/directions';
import verifyDirections from './conditions/directions';
import verifyOffset from './conditions/offset';

function noActionWarning() {
  console.warn('[Actonscroll] No action implemented on user scroll');
}

/**
 * Executes a predefined action when the scroll event is triggered.
 */
export default class Actonscroll {
  /**
   * @param {Object} [options]
   * @param {Document|Element} [options.container]
   * @param {Function} [options.action]
   * @param {Object} [options.conditions]
   * @param {number} [options.throttling]
   * @param {boolean} [options.once]
   */
  constructor(options) {
    this._verifyDirections = verifyDirections.bind(this);
    this._verifyOffset = verifyOffset.bind(this);

    this._init(options);
  }

  _init({
    container = document,
    action = noActionWarning,
    conditions = {},
    throttling = 200,
    once = false,
  } = {}) {
    this._conditions = {};

    this
      .container(container)
      .action(action)
      .conditions(conditions)
      .throttling(throttling)
      .once(once);
  }

  /**
   * Sets the container element within which the scroll will be listened.
   *
   * @example
   * actonscroll.container(document.querySelector('.my-container'));
   *
   * @param {Document|Element} container
   * @returns {Actonscroll}
   */
  container(container) {
    this._container = container;
    this._scrollOffset = getScrollPosition(container);

    return this;
  }

  /**
   * Sets an action to be performed when the scroll event is triggered.
   *
   * @example
   * actonscroll.action(() => console.log('Scrooolling!'));
   *
   * @param {Function} action
   * @returns {Actonscroll}
   */
  action(action) {
    this._action = action;

    return this;
  }

  /**
   * Sets a list of conditions to determine whether the configured action has to be performed or not.
   *
   * @example
   * actonscroll.conditions({
   *   directions: ['up', 'left'],
   *   offset: 200,
   *   custom: () => true,
   * });
   *
   * @param {Object} conditions
   * @param {string[]} conditions.directions Allowed values: <code>'all'</code>, <code>'vertical'</code>, <code>'horizontal'</code>, <code>'up'</code>, <code>'down'</code>, <code>'left'</code>, <code>'right'</code>.
   * @param {number} conditions.offset In pixels.
   * @param {Function} conditions.custom
   * @returns {Actonscroll}
   */
  conditions({ directions, offset, custom }) {
    // Directions
    if (directions && directions.every(direction => directionsConversion[direction])) {
      this._conditions.directions = () => this._verifyDirections(directions);
    }

    // Offset
    if (offset && (!offset.x || offset.x > 0) && (!offset.y || offset.y > 0)) {
      this._conditions.offset = () => this._verifyOffset(offset);
    }

    // Custom
    if (custom) {
      this._conditions.custom = custom;
    }

    return this;
  }

  /**
   * Sets a throttling time (ms) to the scroll event.
   *
   * @example
   * actonscroll.throttling(1000);
   *
   * @param {number} throttling
   * @returns {Actonscroll}
   */
  throttling(throttling) {
    this._throttling = throttling;

    return this;
  }

  /**
   * Determines whether the action should be performed once or not.
   *
   * @example
   * actonscroll.once();
   *
   * @param {boolean} [once]
   * @returns {Actonscroll}
   */
  once(once = true) {
    this._once = once;

    return this;
  }

  /**
   * Executes the configured action after checking that all the conditions are satisfied.
   *
   * @example
   * actonscroll.start();
   */
  start() {
    const runAction = () => {
      const success = this._checkConditions();
      if (success) {
        this._action(success);

        if (this._once) {
          this._stop();
        }
      }
    };

    this._stop();

    this._throttledAction = throttle(runAction, this._throttling);

    this._start();
  }

  _checkConditions() {
    let checkedConditions = {};

    return Object.entries(this._conditions).every(([type, condition]) => {
      checkedConditions[type] = condition();
      return checkedConditions[type];
    }) && checkedConditions;
  }

  /**
   * Removes the listener from the configured container.
   *
   * @example
   * actonscroll.stop();
   */
  stop() {
    this._stop();
  }

  _start() {
    this._container.addEventListener('scroll', this._throttledAction);
  }

  _stop() {
    this._container.removeEventListener('scroll', this._throttledAction);
  }
}
